ReadWriteOncePod access mode for PVs and PVCs

chrishenzie opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR introduces the `ReadWriteOncePod` access mode for PersistentVolumes and PersistentVolumeClaims. This access mode restricts volume access to a single pod on a single node. Please see [KEP#2485](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode) for more context.

This access mode is enforced in two places:

First is during scheduling (see https://github.com/kubernetes/kubernetes/pull/103082). When a pod uses a PVC with ReadWriteOncePod, we check for any other consumers of this PVC in the same namespace. If any are found, consider the pod unschedulable and unresolvable since there can only be one consumer. The failing pod will have error:
```
0/1 nodes are available: 1 node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode.
```

Second is during volume mounting (inside of kubelet). When a volume is mounted (or a block device is mapped), we check the actual state of the world cache to check if the provided volume is already mounted to another pod. If it is, fail the mount operation to prevent access from another pod. The failing pod will have error:
```
MountVolume.SetUp failed for volume "pvc-<ID>" : volume uses the ReadWriteOncePod access mode and is already in use by another pod.
```

#### Special notes for your reviewer:

I tried breaking this into individual chunks in each commit. Hopefully it will make for easier review.

#### Does this PR introduce a user-facing change?
```release-note
Adds the ReadWriteOncePod access mode for PersistentVolumes and PersistentVolumeClaims. Restricts volume access to a single pod on a single node.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode
```

/sig storage
/sig scheduling
/assign @msau42 
/assign @jingxu97

chrishenzie committed:
Validate use of the ReadWriteOncePod access mode

This will only work if the "ReadWriteOncePod" feature gate is enabled.
Additionally, this access mode will only work when used by itself. This
is because when ReadWriteOncePod is used on a PV or PVC, it renders all
other access modes useless since it is most restrictive.

chrishenzie committed:
CSI client helpers for NodeGetCapabilities

chrishenzie committed:
Update helper methods to print and parse ReadWriteOncePod access mode

chrishenzie committed:
Enforce ReadWriteOncePod access mode during mount

chrishenzie committed:
Map PV access modes to CSI access modes

chrishenzie committed:
ReadWriteOncePod PV access mode and feature gate

alculquicondor commented:
It looks like the KEP didn't go through a sig-scheduling review.

The question that comes to mind, regarding the k8s changes: why can't users who need this use pod anti-affinity along with `ReadWriteOnce`?

chrishenzie commented:
@alculquicondor Can we retroactively do that? Any feedback is appreciated.

Users may be able to configure pod anti-affinity rules to produce similar behavior, but I feel like that's working around a lack of feature in Kubernetes that should be implemented. The issue is there is a lack of an access mode that guarantees single writer access (on a single node) to a PersistentVolume. From my understanding, most users assume this is what ReadWriteOnce does, but by definition it doesn't. By adding a new access mode to handle this case, we provide opportunities for users to express with clearer intent how they want their PersistentVolumes to be consumed.

alculquicondor commented:
> but I feel like that's working around a lack of feature in Kubernetes that should be implemented.

If historically PVs are node-oriented, I think the current API is well defined. The use of anti-affinity for more restrictive use cases can be part of the documentation. But perhaps others have a different perspective @msau42 @ahg-g @Huang-Wei ?

Note that adding a new enum value is a breaking change. There is a safe way to do it, please adhere to it: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field

ahg-g commented:
From scheduler perspective, the proposal looks good to me (would certainly have preferred to review the KEP). Having dedicated APIs to support common use cases is cleaner and likely more performant than generic expensive ones like anti-affinity.

chrishenzie commented:
/retest

ehashman commented:
/remove-sig node
/priority important-soon

jingxu97 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1667,6 +1667,9 @@ func ValidatePersistentVolumeSpec(pvSpec *core.PersistentVolumeSpec, pvName stri
+		if utilfeature.DefaultFeatureGate.Enabled(features.ReadWriteOncePod) && mode == core.ReadWriteOncePod {
+			continue
+		}
 		if !supportedAccessModes.Has(string(mode)) {
```
So supportedAccessModes does not include ReadWriteOncePod?

chrishenzie replied:
No, it just includes `ReadWriteOnce`, `ReadOnlyMany`, and `ReadWriteMany`. I plan on adding `ReadWriteOncePod` to this set once the feature is GA.

jingxu97 commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -37,6 +43,10 @@ const Name = "VolumeRestrictions"
 const (
 	// ErrReasonDiskConflict is used for NoDiskConflict predicate error.
 	ErrReasonDiskConflict = "node(s) had no available disk"
+	// ErrReasonUnexpected is used when unexpected errors occur checking volume restrictions.
```
what type of unexpected errors here?

chrishenzie replied:
Not found errors are an example, in which case we should retry to allow time for the PVC to propagate to the lister.

alculquicondor commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1667,6 +1667,9 @@ func ValidatePersistentVolumeSpec(pvSpec *core.PersistentVolumeSpec, pvName stri
 		allErrs = append(allErrs, field.Required(fldPath.Child("accessModes"), ""))
 	}
 	for _, mode := range pvSpec.AccessModes {
+		if utilfeature.DefaultFeatureGate.Enabled(features.ReadWriteOncePod) && mode == core.ReadWriteOncePod {
```
adhere to https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field

chrishenzie replied:
Done, added two commits at the start to introduce validation options for PVs and PVCs, and then extended them in the (currently 4th) commit to check if the feature gate is enabled / if the feature is in use during updates

chrishenzie committed:
Add validation options for PersistentVolumes

These options provide an extensible way of configuring how PVs are
validated

chrishenzie committed:
Add validation options for PersistentVolumeClaims

These options provide an extensible way of configuring how PVCs are
validated

chrishenzie commented:
/assign @alculquicondor

chrishenzie commented:
/assign @jsafrane

alculquicondor commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1638,6 +1639,14 @@ var allowedPVCTemplateObjectMetaFields = map[string]bool{
+// PersistentVolumeValidationOptions contains the different settings for PeristentVolume validation
+type PersistentVolumeValidationOptions struct {
+	// Allow validation of inline PersistentVolume specs
+	ValidateInlinePersistentVolumeSpec bool
```
I think we should keep this structure focused on migrations (safely adding new fields/values).

It looks like `ValidateInlinePersistentVolumeSpec` is not about this, but a different usage of the same API. Did I get this right? If so, please remove this field from here.

I would actually prefer to have to separate functions for `ValidatePersistentVolumeSpec` with/without inline. But that's out of the scope of this PR.

chrishenzie replied:
Yeah that makes sense. Removed the field and re-added the bool arg.

alculquicondor commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1667,12 +1701,15 @@ func ValidatePersistentVolumeSpec(pvSpec *core.PersistentVolumeSpec, pvName stri
 		allErrs = append(allErrs, field.Required(fldPath.Child("accessModes"), ""))
 	}
 	for _, mode := range pvSpec.AccessModes {
+		if opts.AllowReadWriteOncePod && mode == core.ReadWriteOncePod {
```
nit: prefer to use the option to add the value to `supportedAccessModes` (which should be a local variable).

chrishenzie replied:
Done

alculquicondor replied:
you should insert the element once before the loop, regardless of the usage. All that matters is AllowReadWriteOncePod.

alculquicondor commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1965,15 +2002,51 @@ func ValidatePersistentVolumeStatusUpdate(newPv, oldPv *core.PersistentVolume) f
 }
 
+// PersistentVolumeClaimValidationOptions contains the different settings for PersistentVolumeClaim validation
+type PersistentVolumeClaimValidationOptions struct {
```
nit: have a single PersistentVolumeClaimSpecValidationOptions, since you are using it for PVC and PVCTemplate.

We can split into different structs in the future if we see the need, but likely referring to the spec is enough.

chrishenzie replied:
Done. Also renamed PV validation options to `PersistentVolumeSpecValidationOptions` to be consistent (using "Spec").

jsafrane reviewed (commented):
Very well written code! I miss an e2e test, but that can be in a separate PR (and when we have an actual driver...)

chrishenzie commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -86,6 +96,30 @@ func isVolumeConflict(volume *v1.Volume, pod *v1.Pod) bool {
 
+func (pl *VolumeRestrictions) isAccessModeConflict(volume *v1.Volume, pod *v1.Pod) (bool, framework.Code, string) {
+	for _, existingVolume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim != nil && existingVolume.PersistentVolumeClaim != nil {
```
Done, this is easier to read, thank you!

chrishenzie commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1658,10 +1660,28 @@ var supportedVolumeModes = sets.NewString(string(core.PersistentVolumeBlock), st
 	return opts
 }
 
+func accessModesContains(modes []core.PersistentVolumeAccessMode, mode core.PersistentVolumeAccessMode) bool {
```
The existing `AccessModesContains()` is for the external type. Added another helper function for the internal type.

chrishenzie commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -537,12 +537,20 @@ func (og *operationGenerator) GenerateMountVolumeFunc(
 		}
 
+		// Enforce ReadWriteOncePod access mode. This is also enforced during scheduling.
+		if utilfeature.DefaultFeatureGate.Enabled(features.ReadWriteOncePod) &&
```
Correct

chrishenzie commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -86,6 +96,30 @@ func isVolumeConflict(volume *v1.Volume, pod *v1.Pod) bool {
+			// Enforce ReadWriteOncePod access mode. This is also enforced during volume mount in kubelet.
+			if utilfeature.DefaultFeatureGate.Enabled(features.ReadWriteOncePod) && volumeutil.AccessModesContains(existingPVC.Spec.AccessModes, v1.ReadWriteOncePod) {
+				return true, framework.UnschedulableAndUnresolvable, ErrReasonReadWriteOncePodConflict
+			}
```
Yeah this makes sense. I implemented it like so because it seemed like if the access modes contain `ReadWriteOncePod`, then there is the potential for a pod to use it in this fashion. If that is the case, we want to make sure that access mode is supported (e.g. kubelet will enforce this behavior on mount).

chrishenzie replied:
@jsafrane I synced with @msau42 and it was suggested we check this during validation, since use of `ReadWriteOncePod` renders any other included access modes useless. Please see [this commit](https://github.com/kubernetes/kubernetes/pull/102028/commits/6eb291ff81a51f07ca254469d8bcfa16d4ff0635) for the fix. Thoughts on this approach?

jsafrane replied:
As I wrote above, `AccessMode` semantics is not very well defined, esp. when it's an array. Requiring `RWOP` to be the only item there works for me. I commented below, a note about it would be nice. And then we'll see how far it gets during API review.

alculquicondor replied:
@jsafrane did you mean scheduling or API validation? I don't think the scheduler should verify this.

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -110,23 +150,29 @@ func haveOverlap(a1, a2 []string) bool {
 				return framework.NewStatus(framework.Unschedulable, ErrReasonDiskConflict)
 			}
+
+			isAccessModeConflict, code, reason := pl.isAccessModeConflict(v, ev.Pod)
```
I'm not sure if this is the right plugin to do this check. Did you consider volumebinding plugin?

@chendave WDYT? Also, were we checking the other modes before? I couldn't find any usages in scheduler code.

chrishenzie replied:
That plugin could likely also contain this logic if it is a better place. I thought volume restrictions made sense due of [this description of the plugin](https://github.com/kubernetes/kubernetes/blob/v1.21.1/pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go#L108-L110):
```
// It evaluates if a pod can fit due to the volumes it requests, and those that
// are already mounted. If there is already a volume mounted on that node, another pod that uses the same volume
// can't be scheduled there.
```

In this case, a pod would not be schedulable due to the volumes already mounted on that node.

alculquicondor replied:
Actually, it was @cofyc who wrote the volumebinding plugin. Perhaps you have some idea of the questions above?

chrishenzie committed:
Rename access mode contains helper method

So it is consistent with other methods performing the same check (one
for internal and external types)

chrishenzie committed:
Export contains access mode helper method

Will be used during validation of PVs and PVCs

chrishenzie commented on pkg/apis/core/helper/helpers_test.go:
```diff
@@ -100,12 +100,29 @@ func TestGetAccessModesFromString(t *testing.T) {
 	}
 
 	modes = GetAccessModesFromString("RWO,ROX,RWX")
+	if !containsAccessMode(modes, core.ReadWriteOnce) {
```
Done

chrishenzie commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1965,25 +1992,65 @@ func ValidatePersistentVolumeStatusUpdate(newPv, oldPv *core.PersistentVolume) f
+	return opts
+}
+
+func GetValidationOptionsFromPersistentVolumeClaimTemplate(claimTemplate, oldClaimTemplate *core.PersistentVolumeClaimTemplate) PersistentVolumeClaimSpecValidationOptions {
```
Done

chrishenzie commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1648,7 +1656,22 @@ var supportedReclaimPolicy = sets.NewString(string(core.PersistentVolumeReclaimD
 var supportedVolumeModes = sets.NewString(string(core.PersistentVolumeBlock), string(core.PersistentVolumeFilesystem))
 
-func ValidatePersistentVolumeSpec(pvSpec *core.PersistentVolumeSpec, pvName string, validateInlinePersistentVolumeSpec bool, fldPath *field.Path) field.ErrorList {
+func GetValidationOptionsFromPersistentVolume(pv, oldPv *core.PersistentVolume) PersistentVolumeSpecValidationOptions {
```
Done

chrishenzie commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1965,25 +1992,65 @@ func ValidatePersistentVolumeStatusUpdate(newPv, oldPv *core.PersistentVolume) f
+	AllowReadWriteOncePod bool
+}
+
+func GetValidationOptionsFromPersistentVolumeClaim(pvc, oldPvc *core.PersistentVolumeClaim) PersistentVolumeClaimSpecValidationOptions {
```
Done

chrishenzie commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -400,6 +400,7 @@ func (c *csiMountMgr) supportsFSGroup(fsType string, fsGroup *int64, driverPolic
 		klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, access modes not provided"))
 		return false
 	}
+	// TODO: Investigate if this needs updating to support ReadWriteOncePod.
```
@jsafrane @msau42 

Michelle and I synced offline on this. It was suggested that we don't add this to include `ReadWriteOncePod`. If users want to leverage `FSGroupPolicy` with `ReadWriteOncePod` volumes, they will need to create a `CSIDriver` object and set the field. Any thoughts on this?

jsafrane replied:
ack

chrishenzie commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -537,12 +537,20 @@ func (og *operationGenerator) GenerateMountVolumeFunc(
+		// Enforce ReadWriteOncePod access mode. This is also enforced during scheduling.
+		if utilfeature.DefaultFeatureGate.Enabled(features.ReadWriteOncePod) &&
+			actualStateOfWorld.IsVolumeMountedElsewhere(volumeToMount.VolumeName, volumeToMount.PodName) &&
+			util.AccessModesContains(volumeToMount.VolumeSpec.PersistentVolume.Spec.AccessModes, v1.ReadWriteOncePod) {
```
See the above thread for thoughts on this. If we validate ReadWriteOncePod can only be used in isolation then should we still do this check?

chrishenzie commented:
/label api-review

chrishenzie commented:
@alculquicondor and I synced up offline on the scheduler changes.

It looks like the current implementation will not work because we are only marking a single node as unschedulable. It is better if we solve this in the PreFilter phase, update the node cache to map PVCs to the number of pods using them, and then check this cache to enforce the access mode.

To reduce the size of this PR and consolidate the scheduler changes for easier review, I moved them out of this PR and will address them separately.

alculquicondor commented:
and with that, the API bits lgtm (but I'm not an approver yet)

jsafrane commented on pkg/apis/core/types.go:
```diff
@@ -511,6 +511,8 @@ const (
 	ReadOnlyMany PersistentVolumeAccessMode = "ReadOnlyMany"
 	// can be mounted in read/write mode to many hosts
 	ReadWriteMany PersistentVolumeAccessMode = "ReadWriteMany"
+	// can be mounted read/write mode to exactly 1 pod
```
With the new restrictions, IMO a comment here should mention that this mode cannot be used with any other mode.

chrishenzie replied:
Done, added `// cannot be used in combination with other access modes`

alculquicondor commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1666,9 +1688,26 @@ func ValidatePersistentVolumeSpec(pvSpec *core.PersistentVolumeSpec, pvName stri
+	}
+
+	supportedAccessModesCopy := sets.StringKeySet(supportedAccessModes)
+	for _, mode := range pvSpec.AccessModes {
```
I think a single loop would read better

chrishenzie replied:
Done, moved the insertion of `ReadWriteOncePod` into the set outside of the loop / combined the loop logic.

alculquicondor replied:
awesome, lgtm.

Ping @thockin for official api review.

thockin commented:
The linked KEP has no approvers - how is it "implementable" ?

thockin commented:
Since the KEP is merged I'll discuss here.

If I create a PVC with accessMode "ReadWriteOnce" and my cluster uses NFS, I will get a volume that is "ReadWriteMany" capable.  Do we mandate that the PV's accessMode field be scoped down to the PVC's?  IOW, will the PV in this case list "ReadWriteMany" or only "ReadWriteOnce" ?

Assume I have a "ReadWriteOnce" PVC & PV, and it is in-use by pod P1 on node N1, which is now full (no schedulable capacity).  I now create pod P2 using the same PVC, and N2 is the only node with capacity.  Will the scheduler bind P2 to N2 and just let the PVC fail to attach?  Or will scheduler fail to schedule because of the PVC?

It seems unfortunate that this requires a change in all drivers to implement.  Now, I have lost touch with CSI details, but it feels like this SHOULD be a strict subset of ReadWriteOnce support - more like a policy modifier.  If we are smart enough to not schedule 2 pods using it, why does CSI need to know anything at all?

chrishenzie commented:
@thockin 

> The linked KEP has no approvers - how is it "implementable" ?

This was mistakenly left out after first round of KEP review. Opened https://github.com/kubernetes/enhancements/pull/2789 to fix this.

> If I create a PVC with accessMode "ReadWriteOnce" and my cluster uses NFS, I will get a volume that is "ReadWriteMany" capable. Do we mandate that the PV's accessMode field be scoped down to the PVC's? IOW, will the PV in this case list "ReadWriteMany" or only "ReadWriteOnce" ?

No, we don't mandate that the access modes are scoped down for PVs to match the PVC's. When we check if a PV can satisfy a given claim, we check [if the access modes are satisfied](https://github.com/kubernetes/kubernetes/blob/v1.21.2/pkg/controller/volume/persistentvolume/pv_controller.go#L299). So long as [the PV has all of the access modes requested by the PVC](https://github.com/kubernetes/kubernetes/blob/v1.21.2/pkg/controller/volume/persistentvolume/util/util.go#L324), this check will pass.

In other words, you can bind a PVC with `[ReadWriteOnce]` to a PV with `[ReadWriteMany, ReadWriteOnce]`, and the access modes on the PV will not change.

> Assume I have a "ReadWriteOnce" PVC & PV, and it is in-use by pod P1 on node N1, which is now full (no schedulable capacity). I now create pod P2 using the same PVC, and N2 is the only node with capacity. Will the scheduler bind P2 to N2 and just let the PVC fail to attach? Or will scheduler fail to schedule because of the PVC?

In this scenario, P2 will get bound to N2 and let the PVC fail attachment. The scheduler currently does not take these access modes into consideration when making a decision. We could potentially restrict scheduling of pods using `ReadWriteOnce` PVCs to the same host since they will not work anywhere else.

> It seems unfortunate that this requires a change in all drivers to implement. Now, I have lost touch with CSI details, but it feels like this SHOULD be a strict subset of ReadWriteOnce support - more like a policy modifier. If we are smart enough to not schedule 2 pods using it, why does CSI need to know anything at all?

This feature is implementable without any changes to the CSI spec. The reason CSI spec changes are included are to take advantage of new access modes (`SINGLE_NODE_SINGLE_WRITER`, `SINGLE_NODE_MULTI_WRITER`) that disambiguate the number of writers on a single node and to move away from `SINGLE_NODE_WRITER` which has conflicting definitions in the spec. See [this section of the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md#csi-specification-changes) for more context.

chrishenzie commented:
@thockin Clarifying my response to the last question. 

This feature does not require any changes from CSI drivers in order to work. When kubelet talks to the driver, we check if it is capable of supporting the new CSI access modes for single / multi writer. If it is incapable, then when `ReadWriteOncePod` is used, we map it to `SINGLE_NODE_WRITER` (which is what we use for `ReadWriteOnce` and is supported by virtually all CSI drivers). This enables existing (old) CSI drivers to continue working with the new PV access mode.

The reason for introducing these new access modes to the CSI spec is to fix the issue of conflicting definitions over `SINGLE_NODE_WRITER`. Because the spec was not clear about expected behavior when mounting a `SINGLE_NODE_WRITER` volume to multiple target paths (pods), we have a mix of CSI drivers that adhere to the spec and some that don't.

chrishenzie commented:
@thockin When you get a moment to ack I appreciate it. I know we're fast approaching the code freeze and I would like to make sure I have time to address any remaining questions or concerns. If you would like me to schedule time to sync offline on this feature I can do that. Thank you!

thockin commented:
Thanks for the ping.

>> If I create a PVC with accessMode "ReadWriteOnce" and my cluster uses NFS, I will get a volume that is "ReadWriteMany" capable. Do we mandate that the PV's accessMode field be scoped down to the PVC's? IOW, will the PV in this case list "ReadWriteMany" or only "ReadWriteOnce" ?
>
>No, we don't mandate that the access modes are scoped down for PVs to match the PVC's. When we check if a PV can satisfy a given claim, we check [if the access modes are satisfied](https://github.com/kubernetes/kubernetes/blob/v1.21.2/pkg/controller/volume/persistentvolume/pv_controller.go#L299). So long as [the PV has all of the access modes requested by the PVC](https://github.com/kubernetes/kubernetes/blob/v1.21.2/pkg/controller/volume/persistentvolume/util/util.go#L324), this check will pass.

So imagine I have a PV of mode ReadWriteMany. I *REALLY* need a single writer,
so I set PVC mode ReadWriteOncePod.  Could this end up with 2 pods using that
PVC on the same node?

>> Assume I have a "ReadWriteOnce" PVC & PV, and it is in-use by pod P1 on node N1, which is now full (no schedulable capacity). I now create pod P2 using the same PVC, and N2 is the only node with capacity. Will the scheduler bind P2 to N2 and just let the PVC fail to attach? Or will scheduler fail to schedule because of the PVC?
>
>In this scenario, P2 will get bound to N2 and let the PVC fail attachment. The scheduler currently does not take these access modes into consideration when making a decision. We could potentially restrict scheduling of pods using `ReadWriteOnce` PVCs to the same host since they will not work anywhere else.

So, if we're OK with scheduling it and letting it fail in this case, why is
this KEP/PR making scheduler changes?  It seems to me that either we use the
scheduler for both or we use it for neither (my pref :).

Concretely, why not: When a ReadWriteOncePod user lands on a node, we should
mount it in a place that is "private" to the user (e.g. the mountpoint has the
pod UID in the path or something) so if a second pod tries to use the PVC, it
will fail to mount it on its own "private" path.

chrishenzie commented:
> So imagine I have a PV of mode ReadWriteMany. I REALLY need a single writer, so I set PVC mode ReadWriteOncePod. Could this end up with 2 pods using that PVC on the same node?

No, if this happened kubelet would catch it on mount and one pod would start up successfully and the other would fail (due to the ReadWriteOncePod PVC now in use by another pod on this node).

> So, if we're OK with scheduling it and letting it fail in this case, why is this KEP/PR making scheduler changes? It seems to me that either we use the scheduler for both or we use it for neither (my pref :).

The scheduler changes were introduced to fail fast and save kubelet resources managing a pod that is guaranteed to fail.

We cannot rely on just the scheduler because a user can work around this by setting the same node name on both pods. We can rely on just kubelet, but then we will have the same issue of pods perpetually failing. We've been ok with scheduling something known to fail in this situation in the past, but I think enforcing this in the scheduler is a first step towards improving how PV access modes factor into pod assignment so we can save resources downstream.

@jsafrane / @alculquicondor thoughts on this?

alculquicondor commented:
Yes, what Chris describes is what we agreed to from Sig scheduling

On Mon., Jun. 28, 2021, 4:44 p.m. Chris Henzie, ***@***.***>
wrote:

> So imagine I have a PV of mode ReadWriteMany. I REALLY need a single
> writer, so I set PVC mode ReadWriteOncePod. Could this end up with 2 pods
> using that PVC on the same node?
>
> No, if this happened kubelet would catch it on mount and one pod would
> start up successfully and the other would fail (due to the ReadWriteOncePod
> PVC now in use by another pod on this node).
>
> So, if we're OK with scheduling it and letting it fail in this case, why
> is this KEP/PR making scheduler changes? It seems to me that either we use
> the scheduler for both or we use it for neither (my pref :).
>
> The scheduler changes were introduced to fail fast and save kubelet
> resources managing a pod that is guaranteed to fail.
>
> We cannot rely on just the scheduler because a user can work around this
> by setting the same node name on both pods. We can rely on just kubelet,
> but then we will have the same issue of pods perpetually failing. We've
> been ok with scheduling something known to fail in this situation in the
> past, but I think enforcing this in the scheduler is a first step towards
> improving how PV access modes factor into pod assignment so we can save
> resources downstream.
>
> @jsafrane <https://github.com/jsafrane> / @alculquicondor
> <https://github.com/alculquicondor> thoughts on this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/102028#issuecomment-870028175>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ5E6DRLAQPBEBWVPOQWX3TVDNLDANCNFSM445PYGHQ>
> .
>

thockin commented:
Will we make the scheduler ALSO steer ReadWriteOnce pods to the same node, rather than let them schedule on a node that can never work?

IOW if pods p1 and p2 use the same ReadWriteOnce and p1 is on node n1, the ONLY viable place for p2 is n1.  If n1 is not eligible, p2 should fail to schedule.  Right?

chrishenzie commented:
Yes that is the tentative plan for `ReadWriteOnce` in the scheduler

thockin commented:
I'm not a hug fan of the naming (no, I don't have a better one), and I am a bit surprised scheduling wants to take this on, but OK

/approve

ahg-g commented:
> I'm not a hug fan of the naming (no, I don't have a better one), and I am a bit surprised scheduling wants to take this on, but OK
> 
> /approve

With the new scheduling framework, it is fairly easy to add such functionality as a plugin.

chrishenzie commented:
/retest

chrishenzie commented:
/retest

chrishenzie commented:
@jsafrane for lgtm

jsafrane commented:
/lgtm
/approve
for storage side
